### ISSUE TITLE
Fails to decode empty values

### ISSUE THREAD

**Issue by ansermino** [centrifuge/go-substrate-rpc-client#66]
Currently, scale will attempt to decode an empty value and fail with `expected more bytes, but could not decode any more`. This is related to the logic in #64.  

```go
enc := types.MustHexDecodeString("0x00")
var dec types.Call
err := types.DecodeFromBytes(enc, &dec)
if err != nil {
	panic(err)
}
```

Not 100% sure what a good way to resolve this is, happy to put in a fix if I can get some suggestions.

**Comment by ansermino** [comment]
Perhaps this is sufficient: https://github.com/centrifuge/go-substrate-rpc-client/commit/a3a168e3a98b3a15a00cb79391b6bf27c75084bf

**Comment by philipstanislaus** [comment]
`0x00` is not an empty byte array, but rather Scale's encoding for a `None` value of the `Option` type. 

You are getting `0x00` when you encode a n`Option<Call>`, right? 

In that case, you should also decode into a user defined `OptionCall` and handle `0x00` as `None`. An example is here in the codebase: https://github.com/centrifuge/go-substrate-rpc-client/blob/master/types/option_bool.go

**Comment by ansermino** [comment]
Perhaps I've failed to identify the actual problem here.

When querying a storage item of type `Option<T>`, the returned type is `scale(T)` or `0`, not a scale encoded option. The result returned from `getStorageRaw()` is empty when I query a non-existent option type.

I can (and will) handle this case by adding a custom decode method, but this does seem to be a flaw in the storage retrieval mechanism.

**Comment by philipstanislaus** [comment]
@ansermino can you give me an example for that from the Substrate code base? Happy to have a look on Monday.

**Comment by ansermino** [comment]
Here's a snippet you can run against centrifuge-chain. `Authorship.Author` ([source](https://github.com/paritytech/substrate/blob/7c751574d8c53108e78440bfa5dffb80ab644972/frame/authorship/src/lib.rs#L158)) seems to always be none for a dev chain.

```go
api, err := gsrpc.NewSubstrateAPI(config.Default().RPCURL)
if err != nil {
	panic(err)
}

meta, err := api.RPC.State.GetMetadataLatest()
if err != nil {
	panic(err)
}

key, err := types.CreateStorageKey(meta, "Authorship", "Author", signature.TestKeyringPairAlice.PublicKey, nil)
if err != nil {
	panic(err)
}

var c types.AccountID
err = api.RPC.State.GetStorageLatest(key, &c)
if err != nil {
	panic(err)
}
```

Fwiw:
![Screenshot_2020-03-20_18-56-17](https://user-images.githubusercontent.com/14164624/77212173-8d0cd800-6adc-11ea-9e2d-009e6ba17fa7.png)

**Comment by philipstanislaus** [comment]
Thanks @ansermino , I dug a bit deeper into this, and you are right, they just return an empty byte slice if the value is `None`. This is not a SCALE/encoding problem though, but rather a storage interface problem.

I believe we should change `GetStorage`, `GetStorageLatest`, `GetChildStorage` and `GetChildStorageLatest` to return `(empty bool, err error)` instead of just `error`. 

That way you can simply check the first return value for emptiness: 

```go
var author types.AccountID
empty, err := api.RPC.State.GetStorageLatest(key, &author)
if err != nil {
	panic(err)
}
if empty {
	// do something
}
``` 

I think this is no problem for `GetStorageRaw`, `GetStorageRawLatest`, `GetChildStorageRaw`, `GetChildStorageRawLatest`, `SubscribeStorageRaw`, since you can simply check the length of the value.

What do you think?

**Comment by ansermino** [comment]
@philipstanislaus that makes a lot of sense.

My only note would be that returning `ok bool, err error` might be more idiomatic, where `ok` is true for non-empty values.

**Comment by philipstanislaus** [comment]
Sounds good, should be fixed with b0e58dd . Let me know whether that works!
